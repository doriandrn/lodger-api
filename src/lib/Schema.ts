import { RxJsonSchema, RxJsonSchemaTopLevel } from "rxdb"
import SchemaError from './Error'
import { env } from './defs/env'

enum Errors {
  missingProps = 'Missing properties on schema %%',
  propExists = 'Property "%%" already exists',
  idUndef = 'ID for field %% cannot be undefined',
  invalidField = 'Invalid field %% supplied'
}

/**
 * @interface LodgerSchema
 */
interface LodgerSchema extends RxJsonSchema {
  add (id: string, field: Field): void
}

type LodgerSchemaOptions = {}

const datamodelDir = ['dev', 'test']
  .indexOf(env) > -1 ? '.schemas' : '.'

type SchemaProperties<Interface> = {
  [k in keyof Interface] ?: RxJsonSchemaTopLevel
}

/**s
 *
 *
 * @class Schema
 * @extends {RxJsonSchema}
 * @implements {LodgerSchema}
 */
export default class Schema<Name extends string, Interface> implements RxJsonSchema, LodgerSchema {
  readonly type = 'object'
  readonly version = 0
  readonly properties : SchemaProperties<Interface> = {}
  readonly required: string[] = []
  readonly indexes : string[] = []

  /**
   * Constructs a valid RxJsonSchema out of a Lodger Form Data item
   *
   * @param {LodgerFormCreator} form
   * @param {boolean} [addCommonMethods]
   *
   * @memberof Schema
   * @returns {RxJsonSchema} schema
   */
  constructor (
    readonly name: Name,
    fields ?: { [i: string]: Field },
    options?: LodgerSchemaOptions
  ) {
    if (!fields || !Object.keys(fields).length)
      throw new SchemaError(Errors.missingFields, { name })

    Object.keys(fields).map(f => this.add(fields[f].id || f, fields[f]))

    if (options) {}
  }

  /**
   * Adds properties programatically as
   * we also need to fill in the required array
   *
   * @param {FieldCreator} field
   * @memberof Schema
   */
  add (id: string, field ?: Field) {
    if (!id)
      throw new SchemaError(Errors.idUndef, field)

    if (this.properties[id])
      throw new SchemaError(Errors.propExists, id)

    if (field && !field.rxSchema)
      throw new SchemaError(Errors.invalidField, field)

    const { rxSchema, v, storage, _index } = field

    if (storage !== 'db') return

    const required =  v && v.indexOf('required') > -1
    this.properties[id] = rxSchema || {}

    if (required && this.required.indexOf(id) < 0)
      this.required.push(id)

    if (_index) {
      this.indexes.push(id)
    }
  }

  get ids () {
    return Object.keys(this.properties)
  }
}
