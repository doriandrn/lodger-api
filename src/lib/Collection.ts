import Schema, { LodgerSchemaCreator } from './Schema'
import { RxCollectionCreator } from 'rxdb'

class Collection<Name extends string, Interface> extends Schema<Name, Interface> implements RxCollectionCreator {
  constructor (data: LodgerSchemaCreator<Interface>) {
    super(data)
  }
}
