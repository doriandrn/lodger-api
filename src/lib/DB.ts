import * as RxDB from 'rxdb'

import memoryAdapter from 'pouchdb-adapter-memory'
import idbAdapter from 'pouchdb-adapter-idb'
import httpAdapter from 'pouchdb-adapter-http'

import { env } from 'defs/env'

// RxDB.QueryChangeDetector.enable()
// RxDB.QueryChangeDetector.enableDebugging()

// const adapters = {
//   production: ['http', 'idb'],
//   development: ['memory']
// }

// Object.keys(adapters).forEach(env, () => {
//   adapters[env].forEach((adapterType: Adapter) => {
//     if (NODE_ENV !== env) return
//     const adapter = `pouchdb-adapter-${adapterType}`
//     RxDB.plugin(adapter)
//   })
// })

switch (env) {
  default:
    RxDB.plugin(memoryAdapter)
    break

  case 'production':
    RxDB.plugin(httpAdapter)
    RxDB.plugin(idbAdapter)
    break
}

export default RxDB

// /**
//  * RxDB instantiator
//  *
//  * @export
//  * @async
//  * @param {RxDB.RxCollectionCreator[]} collections
//  * @param {RxDB.RxDatabaseCreator} [config]
//  * @returns {RxDB} the fresh database
//  */
// export default async function (
//   collections: RxDB.RxCollectionCreator[],
//   config?: RxDB.RxDatabaseCreator
// ) {
//   debug('Initing')
//   const db =

//   // show leadership in title
//   db.waitForLeadership().then(() => {
//     if (env !== 'dev') return
//     if (!process.browser && env !== 'dev') return
//     document.title = `â™› ${document.title}`
//   })
//   await Promise.all(collections.map(c => db.collection(c)))
//   return db
// }
