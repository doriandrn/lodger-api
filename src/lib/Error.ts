// import fx from 'fx'

/**
 * Error logger
 */
export default class LodgerError extends Error {
  constructor(m: string, details?: any) {
    if (details) {
      m = m.replace('%%', `"${JSON.stringify(details)}"`)
    }
    super(m)

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, LodgerError.prototype)
  }
}
