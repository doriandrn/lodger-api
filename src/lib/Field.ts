import { RxJsonSchemaTopLevel, RxDocument, JsonSchemaTypes } from "rxdb";
import FieldError from './Error'

// These 3 lines are a hack for Rollup & Jest to work together.
// Test with: yarn rollup -c & jest field
import { strings, numbers, arrays, objects } from './String'
import S from './String'

// This should only be imported for DEVELOPMENT PURPOSES
import fakeData from 'helper/dev/fakeData'

import { computed } from 'mobx'

const { String } = S

type ItemExcludableFrom = 'db' | 'addForm' | 'editForm' | 'all'
type ReferenceTaxonomy = Plural<Taxonomie>

// Can be used to pass in the field additional params for when calculating the value
type FieldGivenContext<T> = {
  selectedDoc ?: RxDocument<T>,
  activeDoc : RxDocument<T>
}

type FieldTypes = keyof typeof strings |
  keyof typeof numbers |
  keyof typeof arrays |
  keyof typeof objects |
  undefined

// enum cheiImutabile { 'primary' | 'index' | 'encrypted' | 'required'

declare global {
  // type ID<X extends Taxonomie> = string

  type FieldCreator = {
    // id : keyof T // item's identifier, correlates to DB's item key

    placeholder ?: string // sample data

    // values
    default?: any | Function
    value?: (context ?: FieldGivenContext<any>) => any

    // field description
    type ?: FieldTypes // our form types. DEFAULT: 'string'
    required ?: boolean // if the field is required
    encrypted ?: boolean // encrypt field's value
    index ?: boolean // should be indexable <=> sorts & search
    primary ?: boolean // skip _id field and make this primary instead
    // excludeFrom ?: ItemExcludableFrom | ItemExcludableFrom[]

    key ?: string // or name -> to be used for grouping checkboxes / radio groups
    options ?: Array<any> | Object // for selects and other multiple options els
    fieldset ?: number // corresponding fieldsets' array number - groups fields into fieldsets

    // numeric inputs
    step ?: number // incrementer for number inputs, step or multipleOf
    min ?: number
    max ?: number
    final ?: boolean // cannot be modified at all
    freezed ?: boolean // cannot be changed later on after first init

    v ?: string // validation string (for vee-validate)

    // references
    ref ?: ReferenceTaxonomy // referenceTaxonomy
    indexRef ?: boolean // index by references?

    // frontend stuff
    // showInList ?: 'primary' | 'secondary' | 'details'[]
    preview ?: number
    onclick ?: {
      [method: string]: string
    }
    oninput ?: {
      [method: string]: string
    }
    focus ?: boolean // if this should be the first option to focus on
    search ?: boolean // if searchable
    items ?: { type: 'string' | 'object' | 'array', properties: { [k: string]: any} }
  }
}

/**
 * @interface FieldAPI
 */
interface FieldAPI {
  readonly default : any | Function
  readonly rxSchema : RxJsonSchemaTopLevel
  readonly _type: FieldTypes

  value (context ?: FieldGivenContext<any>): any
  onclick ?: (context ?: FieldGivenContext<any>) => void
}

// excluded _id, we use it as a key. others are optional
export type FieldsCreator<Schema> = {
  [i in Exclude<keyof Schema, ['_id', '@', 'upd@']>]: FieldCreator
}


/**
 *
 * @class Form Field Item
 * @implements {SchemaField}
 * @requires [String]
 * @extends RxJsonSchemaTopLevel
 */
export class Field implements FieldAPI {
  readonly type: JsonSchemaTypes = 'string'

  // readonly label : string = translate(_id, 'fields')

  readonly ref ?: ReferenceTaxonomy
  readonly items ?: { type: 'string' | 'object' | 'array', properties: { [k: string]: any} }
  readonly multipleOf ?: number // multiplier if number
  readonly v ?: string // validation string

  readonly storage ?: 'db' | 'store'  = 'db' // where to store data, in db or store

  readonly default : any
  readonly value : (context ?: FieldGivenContext<any>) => any = () => this.default || undefined
  readonly fakeValue : any
  readonly preview ?: number
  readonly oninput ?: {
    transform ?: string
  }

  readonly _type ?: FieldTypes
  readonly _index ?: boolean // should be indexed to search for
  readonly encrypted ?: boolean
  key : string = 'index'
  focus ?: boolean
  search ?: boolean
  fieldset ?: number
  final ?: boolean
  readonly options ?: string[] | {}
  freezed ?: boolean // cannott be edited by the user after init

  /**
   * Creates an instance of Field.
   *
   * @param {FieldCreator<T>} data
   * @memberof Field
   */
  constructor (
    data ?: FieldCreator
  ) {
    if (!data) {
      // throw new FieldError('Field could not be created. No data supplied.')
      return
    }

    const {
      ref,
      focus,
      encrypted,
      index,
      indexRef,
      type,
      step,
      required,
      v,
      value,
      preview,
      oninput,
      key,
      fieldset,
      freezed,
      final,
      search,
      options
    } = data

    if (preview !== undefined) this.preview = preview
    this._type = type // hold this for reference
    this.type = String(type || '').asRxDBType
    if (oninput) this.oninput = oninput
    if (focus) this.focus = focus
    if (options) this.options = options
    if (final) this.final = final
    if (key) this.key = key
    if (key === 'attachments') this.type = undefined
    if (search) this.search = search
    if (index !== undefined) this._index = true
    if (fieldset !== undefined) this.fieldset = fieldset
    if (encrypted !== undefined) this.encrypted = encrypted

    // transform the ref
    if (ref) {
      this.ref = ref
      this.items = { type: 'string' }
      if (indexRef) this.index = indexRef
    }

    // steps for numbers
    if (step !== undefined) {
      if (this.type !== 'number')
        throw new FieldError('Type must be "number" for .step')

        this.multipleOf = step
    }

    // hook in required to validation string
    if (required) {
      this.v = required && v && v.indexOf('required') < 0 ?
        v :
        `required|${v}`
    }

    if (freezed)
      this.freezed = true

    // assign default value, can be undefined
    if (data.default !== undefined)
      this.default = data.default

    // bind the value function
    const { storage } = this
    if (value && typeof value === 'function')
      this.value = value.bind({ storage })

    Object.defineProperty(this, 'fakeValue', {
      get () { return fakeData[type || 'string'] }
    })

    if (this._type === '$') {
      this.freezed = true

      if (!this.default)
        this.default = ({ displayCurrency }) => ({
          value: 0,
          moneda: displayCurrency
        })

      this.items = {
        type: 'object',
        properties: {
          value: {
            type: 'number'
          },
          moneda: {
            type: 'number'
          }
        }
      }
    }
  }

  @computed get label (): () => string {
    return (o: {[i: string]: string} = {[this.key]: 'unnamed'}) => {
      try {
        return o[this.key]
      } catch (e) {
        return 'undefined label'
      }
    }
  }


  /**
   * Used for Schema constructors,
   * returns only the properties needed for it
   */
  get rxSchema (): RxJsonSchemaTopLevel {
    const schema = {}
    const excludes = ['storage', 'value', 'v', 'default']
    Object.keys(this).forEach(prop => {
      if (this[prop] === undefined) return
      if (excludes.indexOf(prop) > -1) return

      schema[prop] = this[prop]
    })
    return schema
  }
}
