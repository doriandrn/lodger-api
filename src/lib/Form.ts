/**
 * Forms for Lodger
 * are quite diferrently structured
 * than a normal JsonSchema
 */
import Debug from 'debug'
import { RxCollectionCreator } from 'rxdb'

import { env } from './defs/env'
import FormError from './Error'
import { GetterTree } from 'vuex'
import { RootState } from './Store'

type ItemReference = Plural<Taxonomie> | object
type FormExcludables = 'db' | 'addForm' | 'editForm' | 'all'
type ItemExcludableFrom = FormExcludables[]

export type FieldCreator = {
  id: string,
  name?: string,

  label?: string
  placeholder?: string

  type?: FormItemTypes
  required?: boolean
  encrypted?: boolean

  default?: any
  value?: any

  step?: number,
  index?: boolean,
  ref?: ItemReference
  items?: object
  indexRef?: boolean

  excludeFrom?: ItemExcludableFrom

  v?: string // validation string
  click?: string
  showInList?: 'primary' | 'secondary' | 'details'[]
}



/**
 * Errors Definition
 * @readonly
 * @enum {string}
 *
 * @todo account for translations
 */
enum Errors {
  invalidRequested = 'Invalid form requested: %%',
  invalidName = 'Invalid name supplied',
  noData = 'Form %% is missing data',
  missingName = 'Forms should have a name',
  missingPlural = 'A plural definition is required for %%'
}

if (env === 'test')
  Debug.enable('Form:*')

export type LodgerFormCreator = {
  name?: string
  plural: Plural<Taxonomie>
  fields: FieldCreator[]

  methods?: { [k: string]: () => void }
  statics?: { [k: string]: () => void }
  sync?: boolean
  setari?: any
}

const formsPath = ['dev', 'test'].indexOf(env) > -1 ? 'forms' : '.'


// export interface LodgerFormConstructor {
//   new (data: LodgerFormCreator): LodgerForm
// }

interface LodgerForm {
  name: string
  collection: undefined | RxCollectionCreator
  indexables ?: string[]
}

/**
 * Forms are read from within the `lib/forms/` directory
 *
 * @class Form
 * @implements {LodgerForm}
 */
class Form implements LodgerForm {
  name: string
  fields: FieldCreator[]
  collection: undefined | RxCollectionCreator

  readonly indexables ?: string[]
  readonly plural : Plural<Taxonomie>

  /**
   * Creates an instance of Form.
   *
   * @param {LodgerFormCreator} data - Form input data
   * @param {boolean} [generateRxCollection=true] - some forms don't require this
   *
   * @memberof Form
   */
  constructor (
    data: LodgerFormCreator,
    generateRxCollection : boolean = true
  ) {
    const { fields, name, plural, methods, statics } = data
    if (!name) throw new FormError('Form should have a name %%', data)
    if (!fields || !Object.keys(fields).length)
      throw new FormError('missing fields on form %%', name)

    this.name = name
    this.plural = plural
    this.fields = fields

    if (generateRxCollection) {
      const schema = new Schema(data, true)
      const collection = {
        name: plural,
        schema,
        methods,
        statics
      }
      this.collection = collection
      this.indexables = Object.keys(schema.properties).filter(prop => schema.properties[prop].index)
    }
  }

  /**
   * Makes a Vue-ready $data {object} suitable to be completed
   * by the user in the end form
   * as it will turn reactive
   *
   * @summary for new forms, values are all undefined
   *
   * @param {Boolean} isNewForm - mostly used for 'add' forms
   * @returns {Object} data item (Vue $data - ready)
   */
  value (
    isNewForm: boolean,
    getters?: GetterTree<any, RootState>
  ): FormValue {
    const { fields, name } = this
    // const debug = Debug('Form:value')
    let $data = {} as any

    fields.forEach(camp => {
      const { label, required, click, excludeFrom } = camp
      let { id, value } = camp

      let _def = camp.default

      if (click && !id) camp.id = click

      // skip excluded fields
      if (isNewForm && excludeFrom && (excludeFrom.indexOf('db') || excludeFrom.indexOf('addForm')))
        value = undefined

      // apply getters to funcs
      value = typeof value === 'function' && getters ? value(getters) : undefined
      _def = typeof _def === 'function' ? _def(getters) : undefined

      // label
      camp.label = label || `${name ? `${name}.new.` : ''}${id}`

      // validarea de required
      if (required || (camp.v && camp.v.indexOf('required') < 0))
        camp.v = `required|${camp.v || ''}`

      // valoarea finala
      $data[id] = null
      $data[id] = value !== null && value !== undefined ? value : _def
    })

    return this.handleOnSubmit($data)
  }

  /**
   * Manipulates the final data before submitting the form to the DB
   *
   * @param data
   */
  handleOnSubmit (
    data : FormValue
  ) {
    const manipulatedData: any = {}

    // not data.denumire pt servicii :/
    if (!data.la && !data.denumire) data.la = Date.now()
    Object.keys(data).forEach(what => {
      let value = data[what]
      if (value === null || value === 'undefined') {
        return
      }

      manipulatedData[what] = value
    })

    if (!context) return manipulatedData
    const { referencesIds } = context

    Object.assign(manipulatedData, referencesIds)

    return manipulatedData
  }

  /**
   * Loads a 'known' form by name
   *
   * @param name
   */
  static async load (name: string): Promise<Form> {
    const debug = Debug('lodger:Form')
    if (!name)
      throw new FormError('no name supplied for form')

    const formPath: string = `${formsPath}/${String(name).toLowerCase()}`

    try {
      const formData: LodgerFormCreator = await import(formPath)
      Object.assign(formData, { name })
      debug('âœ“', name)
      return new Form(formData)
    } catch (err) {
      debug('x', name)
      throw new FormError(err)
    }

  }
}

export {
  Form,
  Errors,
  prepareRxSchema
}
