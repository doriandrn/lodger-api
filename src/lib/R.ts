import Vue from 'vue'
import { getRxDocumentById } from './helpers/subscribers'

/**
 * Main holder for temporary items subscribed to
 * RENDERLESS Vue Component
 *
 * -> a vue helper for reactivity
 * holds RX documents
 * and methods to accezss / manipulate them
 *
 * @export
 * @returns {Vue} data holder object
 */
export default new Vue({
  data () { return {
    subsData: {}
  }},

  computed: {
    /**
     * Itmes ids of a requested taxonomy by it's plural
     *
     * @param {Plural<Taxonomie>} taxonomy
     * @param {subscriberName} subscriberName
     *
     * @returns {ItemID[]} the IDs of contained items
     */
    ids () {
      return (taxonomy: Plural<Taxonomie>, subName: string) => {
        return Object.keys(this.subsData[subName][taxonomy])
      }
    }
  },
  methods: {
    /**
     * Gets an item from existing temporary items
     * or looks it up in the DB
     *
     * @param {Plural<Taxonomie>} taxonomie
     * @param {string} id
     * @param {string} [subscriberName]
     * @returns {object} the item
     *
     * @todo create a type for the returned item
     */
    async getItem (
      taxonomie: Plural<Taxonomie>,
      id: string,
      subscriberName ?: string
    ) {
      let item: RxDocument<Taxonomie> | undefined
      const debug = Debug('lodger:getItem')

      if (subscriberName === undefined)
        subscriberName = 'main'

      const { subsData } = this

      try {
        const s = subsData[subscriberName][taxonomie]
        if (s.docs && s.docs.length) return getRxDocumentById(s.docs, id)
      } catch (e) {
        Object.keys(this.subsData).forEach(sub => {
          if (item) return
          const s = subsData[sub][taxonomie]
          if (!(s && s.docs && s.docs.length)) return
          item = getRxDocumentById(s.docs, id)
          if (item) debug('item gasit din a 2a', { taxonomie, subscriberName, s, item })
        })

      } finally {
        // should never get here maybe because data displayed should be available.
        // item = await collections[plural].findOne(id).exec()
      }
      return item
    }
  }
})
