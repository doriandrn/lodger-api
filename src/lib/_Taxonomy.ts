import { predefinite } from 'forms/serviciu'


// /**
//  * Pt taxonomia ceruta
//  * ia formul
//  * si tot ce are nevoie de Id de altceva
//  * se populeaza
//  *
//  * @param {Object} { references, getters }
//  * @returns {Object} eg { asociatieId: 'XXXX' }
//  */
// function assignRefIdsFromStore (context: any) {
//   const { references, getters } = context
//   if (!(references && references.length)) return

//   const refsObj = {}

//   references.map((tax: Taxonomie) => {
//     refsObj[`${tax}Id`] = getters[`${tax}/selected`]
//   })

//   return refsObj
// }


// const xx: LodgerTaxonomyCreator<Taxonomie> = new Taxonomy('asociatie', 'form', 'collection')
/**
 * gets the sorting options for tax
 * @returns an object with each key used as a sorting option
 */
// get sortOptions () {
//   const { indexables, name } = this

//   if (!['serviciu', 'contor'].indexOf(name)) {
//     indexables.push('la')
//   }

//   // TODO: !!! ia din common methods
//   const sorts = {}
//   indexables.forEach(indexable => {
//     const label = `sort.${indexable === 'name' ? 'az' : indexable}`
//     Object.assign(sorts, { [indexable]: { label } })
//   })

//   // debug(`${name} => sortable fields`, sorts)

//   return sorts
// }
/**
 * Items to be display to user,
 * @returns {Object} the keys of the fields: their position
 *
 */
// get __displayItemKeys () {
//   const { fields } = this.data

//   return Object.assign({}, ...fields
//     .filter(field => field.showInList)
//     .map(field => ({ [field.id]: field.showInList }) )
//   )
// }


// // Filters the documents array for the one with the id
// export const getRxDocumentById = (docs: RxDocument<Taxonomie, any>[], id: string) => {
//   if (!docs.length)
//     throw new LodgerError('Empty docs provided: %%')
//   const doc = docs.filter(doc => doc._id === id)[0]
//   if (!(doc && isRxDocument(doc)))
//     throw new LodgerError('No document found %%', { id })
//   return doc
// }

//
  /**
   * Active document for taxonomy
  */
//  protected set _activeDocument (docHolder: ActiveDocumentHolder) {
//   let { taxonomie, doc } = docHolder
//   const debug = Debug('lodger:_activeDocument')
//   const gName = `${taxonomie}/activeDoc`
//   const { store } = this

//   if (!store.getters.hasOwnProperty(gName)) {
//     Object.defineProperty(store.getters, gName, {
//       configurable: false,
//       get () { return doc },
//       set (newDoc) { doc = newDoc }
//     })
//   } else {
//     store.getters[gName] = doc
//   }
// }
