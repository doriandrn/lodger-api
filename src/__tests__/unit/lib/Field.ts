import { Field } from '~/lib/Field'

/**
 * Setup a dummy id to use everywhere
 */
const id = 'test'

describe('Field', () => {
  describe('new()', () => {
    describe('+', () => {
      test('constructs with ID', () => {
        const field = new Field({ id })
        console.info('Field', field)
        expect(field).toBeDefined()
        expect(field.id).toEqual(id)
        expect(field.type).toBe('string')
      })
    })
  })

  describe('.ref - references', () => {
    describe('+', () => {
      test('is undefined if no references passed', () => {
        const field = new Field({ id })
        expect(field.ref).toBeUndefined()
      })

      test('adds references required stuff', () => {
        const ref = 'asociatii'
        const field = new Field({ id, ref })

        expect(field.ref).toBe(ref)
        expect(field.items.type).toBe('string')
      })
    })
  })

  describe('.type', () => {
    describe('+', () => {
      test('gets converted', () => {
        const field = new Field({ id, type: 'dateTime' })
        expect(field.type).toBe('number')
      })
    })
  })

  describe('.multipleOf', () => {
    describe('+', () => {
      test('transforms .step to .multipleOf if type is "number"', () => {
        const step = 5
        const field = new Field({ id, type: 'number', step })
        expect(field.multipleOf).toBe(step)
      })
    })

    describe('-', () => {
      test('throws if type is not "number"', () => {
        expect(() => new Field({ id, step: 4 })).toThrow()
      })
    })
  })

  describe('.default', () => {
    describe('+', () => {
      test('assigns a static value', () => {
        const defVal = 'foo'
        const field = new Field({ id, default: defVal })
        expect(field.default).toEqual(defVal)
      })

      test('if function calculates its value', () => {
        const defVal = 'foo'
        const defFunc = () => defVal
        const field = new Field({ id, default: defFunc })
        expect(field.default).toEqual(defVal)
      })
    })
  })

  describe('.value()', () => {
    describe('+', () => {
      test('ok without context', () => {
        const field = new Field({ id, value: () => 'foo' })
        expect(field.value()).toEqual('foo')
      })

      test('applies context given data', () => {
        const getterTree = {
          foo: 'bar'
        }
        const field = new Field({ id, value: ({ getters }) => getters['foo'] })
        expect(field.value({ getters: getterTree as any })).toEqual('bar')
      })

      test('is undefined if "value" property missing', () => {
        const field = new Field({ id })
        expect(field.value()).toBeUndefined()
      })
    })

    describe('-', () => {

    })
  })

})
