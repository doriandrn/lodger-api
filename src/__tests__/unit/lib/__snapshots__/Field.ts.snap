// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.rxSchema + matches snapshot: testField 1`] = `
Object {
  "type": "string",
}
`;
